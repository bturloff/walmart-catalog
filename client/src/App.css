nav {
  background-color: #041e42;
}

nav .brand-logo img {
  height: -webkit-fill-available;
}

.search-wrapper {
  background-color: blue;
}

#nav-mobile > li {
  /* margin-right: 30px; */
}
