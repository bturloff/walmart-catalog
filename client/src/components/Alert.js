import React from "react";

/**
 * Renders an alert, letting the user know the search query was invalid
 */

const Alert = () => {
  return (
    <div className="col s12 alert">
      <div className="card yellow lighten-3">
        <div className="card-title">
          <i className="medium material-icons">priority_high</i>
          <span>Sorry No Products match your search</span>
        </div>
        <div className="card-content">
          <ul>
            <li>Check your spelling</li>
            <li>Use different keywords and try again</li>
            <li>Contact your local store</li>
          </ul>
        </div>
      </div>
    </div>
  );
};

export default Alert;
