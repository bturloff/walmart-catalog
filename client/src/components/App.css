.outter-wrapper {
  background-color: #f0f0f0;
  min-height: 100vh;
}

nav {
  background-color: #041e42;
  margin-bottom: 40px;
}

nav .brand-logo img {
  height: -webkit-fill-available;
}

.main {
  margin: 0;
}

.card {
  background-color: #378091;
}

.card-image img {
  height: 250px;
}

.card-content p {
  height: 200px;
  overflow: hidden;
}
.card-title {
  margin: 0;
  padding: 24px;
  height: 120px;
  overflow: hidden;
}

.alert .card-title {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.alert ul:not(.browser-default) > li {
  margin-left: 40px;
  list-style-type: disc;
}

.input-field {
  display: flex;
}

.input-field i {
  position: absolute;
  right: 0;
  color: #4caf50;
  cursor: pointer;
}
