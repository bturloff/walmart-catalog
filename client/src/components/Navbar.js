import React from "react";

function Navbar(props) {
  return (
    <nav>
      <div className="nav-wrapper static">
        <a href="*" className="brand-logo">
          <img src="walmartlogo.png" alt="" />
        </a>
        <ul id="nav-mobile" className="right hide-on-med-and-down">
          <li>
            <a href="*">
              <i className="large material-icons">account_circle</i>
            </a>
          </li>
          <li>
            <a href="*">
              <i className="large material-icons">location_on</i>
            </a>
          </li>
          <li>
            <a href="*">
              <i className="large material-icons">shopping_cart</i>
            </a>
          </li>
        </ul>
      </div>
    </nav>
  );
}
export default Navbar;
