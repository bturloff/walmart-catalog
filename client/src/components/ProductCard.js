import React from "react";
import PropTypes from "prop-types";

const ProductCard = ({ product }) => {
  return (
    <div className="col s12 m4" key={product.itemId}>
      <div className="card hoverable">
        <div className="card-image">
          <img
            className="materialboxed"
            src={product.largeImage}
            alt={product.name}
          />
        </div>
        <div className="card-title">
          <p>{product.name}</p>
        </div>
        <div className="card-content">
          <p>{product.shortDescription}</p>
        </div>
        <div className="card-action">
          <a href="#">Add to Cart</a>
        </div>
      </div>
    </div>
  );
};

ProductCard.propTypes = {
  product: PropTypes.object
};

export default ProductCard;
