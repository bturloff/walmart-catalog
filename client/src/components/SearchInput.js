import React, { Component } from "react";
import Axios from "axios";
import PropTypes from "prop-types";

function SearchInput({ handleKeyDown, handleUpdate, value, handleSubmit }) {
  return (
    <div className="input-field">
      <input
        id="search"
        type="text"
        className="validate"
        placeholder="i.e. ozark sleep patio apple"
        value={value}
        onChange={handleUpdate}
        onKeyDown={handleKeyDown}
      />
      <label htmlFor="search">Search</label>
      <i onClick={handleSubmit} className="small material-icons right">
        search
      </i>
    </div>
  );
}
SearchInput.propTypes = {
  handleKeyDown: PropTypes.func,
  handleUpdate: PropTypes.func,
  handleSubmit: PropTypes.func,
  value: PropTypes.string
};

export default SearchInput;
