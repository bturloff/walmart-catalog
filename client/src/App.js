import React from "react";
import "./App.css";
import Navbar from "./Navbar";

class App extends React.Component {
  state = { searchField: "" };

  _handleKeyDown = e => {
    if (e.key === "Enter") {
      console.log("do validate");
    }
  };

  _handleChange = e => {
    this.setState({ searchField: e.target.value });
  };

  render() {
    return (
      <div>
        <Navbar />
        <div className="container">
          <div className="input-field">
            <input
              id="search"
              type="text"
              className="validate"
              value={this.state.searchField}
              onChange={this._handleChange}
              onKeyDown={this._handleKeyDown}
            />
            <label htmlFor="search">Search</label>
          </div>

          <div className="row">
            <div className="col s12 m6">
              <div className="card blue-grey darken-1">
                <div className="card-content white-text">
                  <span className="card-title">Card Title</span>
                  <p>
                    I am a very simple card. I am good at containing small bits
                    of information. I am convenient because I require little
                    markup to use effectively.
                  </p>
                </div>
                <div className="card-action">
                  <a href="#">This is a link</a>
                  <a href="#">This is a link</a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default App;
