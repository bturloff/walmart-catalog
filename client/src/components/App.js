import React from "react";
import "./App.css";
import Navbar from "./Navbar";
import Axios from "axios";
import SearchInput from "./SearchInput";
import Alert from "./Alert";
import ProductCard from "./ProductCard";

const url = "http://localhost:3001/products";

class App extends React.Component {
  state = { searchField: "", products: [], didSearch: false };

  componentDidUpdate() {
    /** initialize image zoom logic on all elements with 'materialboxed' class */
    var elems = document.querySelectorAll(".materialboxed");
    window.M.Materialbox.init(elems, {});
  }

  /**
   * Listening for "Enter" Key Stroke. Submit when pressed
   */
  _handleKeyDown = e => {
    if (e.key === "Enter") {
      this._handleSubmit();
    }
  };

  /**
   * Handle search submission. Calls Search API for relavent product data
   */
  _handleSubmit = () => {
    const searchField = this.state.searchField;

    /** Search field cannot be empty */
    if (searchField.length === 0) {
      this.setState({ didSearch: true, products: [] });
      return;
    }

    /** Search field is valid */
    Axios.get(`${url}/${searchField}`)
      .then(res => {
        console.log("res", res);
        this.setState({ products: res.data });
      })
      .catch(err => {
        this.setState({ error: err });
      })
      .finally(() => {
        this.setState({ didSearch: true });
      });
  };

  /**
   * Handle updating the searchField state on every keystroke
   */
  _handleUpdate = e => {
    this.setState({ searchField: e.target.value });
  };

  render() {
    const { products, didSearch } = this.state;

    return (
      <div className="outter-wrapper">
        <Navbar />

        <div className="container">
          {/* Search Input */}
          <SearchInput
            handleKeyDown={this._handleKeyDown}
            handleUpdate={this._handleUpdate}
            value={this.state.searchField}
            handleSubmit={this._handleSubmit}
          />

          {/* Main Content */}
          <div className="main row">
            {!didSearch ? (
              <div>Please enter a keyword to search</div>
            ) : products.length === 0 ? (
              <Alert />
            ) : (
              products.map(p => {
                return <ProductCard product={p} key={p.itemId} />;
              })
            )}
          </div>
        </div>
      </div>
    );
  }
}

export default App;
