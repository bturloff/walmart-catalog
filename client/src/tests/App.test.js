import React from "react";
import { shallow } from "enzyme";
import App from "../components/App";
import Navbar from "../components/Navbar";
import ProductCard from "../components/ProductCard";
import products from "../../../server/src/products.json";
import Alert from "../components/Alert";
import SearchInput from "../components/SearchInput";

test("App renders without crashing", () => {
  shallow(<App />);
});

/**
 * Navbar
 */
test("Navbar should match snapshot", () => {
  const navbar = shallow(<Navbar />);
  expect(navbar).toMatchSnapshot();
});

/**
 * ProductCard
 */
test("ProductCard renders with one produce", () => {
  const mock = products[0];
  const productCard = shallow(<ProductCard product={mock} />);
  expect(productCard.contains(mock.name)).toBeTruthy();
  expect(productCard.contains(mock.shortDescription)).toBeTruthy();
  expect(productCard).toMatchSnapshot();
});

/**
 * Alert
 */
test("Alert renders correctly", () => {
  const alert = shallow(<Alert />);
  expect(alert).toMatchSnapshot();
});

/**
 * SearchInput
 */
test("SearchInput renders correctly", () => {
  const searchInput = shallow(<SearchInput value="backpack" />);
  expect(searchInput.contains("backback"));
  expect(searchInput).toMatchSnapshot();
});
test("SearchInput updates correctly", () => {
  const handleKeyDown = jest.fn();
  const handleUpdate = jest.fn();
  const searchInput = shallow(
    <SearchInput handleKeyDown={handleKeyDown} handleUpdate={handleUpdate} />
  );
  searchInput.find("input").simulate("change");
  expect(handleUpdate).toBeCalled();
  searchInput.find("input").simulate("keyDown", { keyCode: 40 });
  expect(handleKeyDown).toBeCalled();
});
test("SearchInput submits on click", () => {
  const handleSubmit = jest.fn();
  const searchInput = shallow(<SearchInput handleSubmit={handleSubmit} />);
  searchInput.find("i").simulate("click");
  expect(handleSubmit).toBeCalled();
});

/**
 * App
 */
test("App renders correctly", () => {
  const mock = products;
  const app = shallow(<App />);
  app.instance().componentDidUpdate = jest.fn();

  /** message renders when user has not searched yet */
  app.instance().setState({ products: [], searchField: "", didSearch: false });
  expect(app.contains("Please enter a keyword to search")).toBeTruthy();

  /** Alert appears when user submits invalid search field */
  app.instance().setState({ products: [], searchField: "", didSearch: true });
  expect(app.find(Alert)).toHaveLength(1);

  /** Appropriate number of Product Cards render based on num products in state */
  app.instance().setState({ products: mock, searchField: "", didSearch: true });
  expect(app.find(ProductCard)).toHaveLength(8);
});
