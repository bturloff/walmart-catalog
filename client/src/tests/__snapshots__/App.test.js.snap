// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert renders correctly 1`] = `ShallowWrapper {}`;

exports[`Navbar should match snapshot 1`] = `ShallowWrapper {}`;

exports[`ProductCard renders with one produce 1`] = `ShallowWrapper {}`;

exports[`SearchInput renders correctly 1`] = `ShallowWrapper {}`;
